src/particles/beam/BeamParticleContainerInit.cpp

```diff
@@ -60,190 +60,243 @@ namespace
+                ParticleUtil::get_position_unit_cell(r, ppc, i_part, engine, rand_ppc);
 
                 amrex::Real x = plo[0] + (i + r[0])*dx[0];
                 amrex::Real y = plo[1] + (j + r[1])*dx[1];
-                amrex::Real z = plo[2] + (k + r[2])*dx[2];
+                amrex::Real z = plo[2] + (islice + r[2])*dx[2];
 
                 if (rand_ppc[0] + rand_ppc[1] + rand_ppc[2] == false) {
                     // If particles are evenly spaced, discard particles
                     // individually if they are out of bounds
-                    if (z >= a_zmax || z < a_zmin ||
-                        ((x-x_mean)*(x-x_mean)+(y-y_mean)*(y-y_mean)) > a_radius*a_radius) {
+                    if (z >= z_max || z < z_min ||
+                        ((x-x_mean)*(x-x_mean)+(y-y_mean)*(y-y_mean)) > radius*radius) {
                             continue;
                         }
                 } else {
                     // If particles are randomly spaced, discard particles
                     // if the cell is outside the domain
                     amrex::Real xc = plo[0]+i*dx[0];
                     amrex::Real yc = plo[1]+j*dx[1];
-                    amrex::Real zc = plo[2]+k*dx[2];
-                    if (zc >= a_zmax || zc < a_zmin ||
-                        ((xc-x_mean)*(xc-x_mean)+(yc-y_mean)*(yc-y_mean)) > a_radius*a_radius) {
+                    amrex::Real zc = plo[2]+islice*dx[2];
+                    if (zc >= z_max || zc < z_min ||
+                        ((xc-x_mean)*(xc-x_mean)+(yc-y_mean)*(yc-y_mean)) > radius*radius) {
                             continue;
                         }
                 }
 
                 const amrex::Real density = get_density(x, y, z);
-                if (density < a_min_density) continue;
+                if (density < min_density) continue;
 
                 amrex::Real u[3] = {0.,0.,0.};
-                get_momentum(u[0],u[1],u[2], engine);
+                get_momentum(u[0], u[1], u[2], engine);
 
                 const amrex::Real weight = density * scale_fac;
-                AddOneBeamParticle(rarrdata, iarrdata, x, y, z, u[0], u[1], u[2], weight,
-                                   pid, pidx, phys_const.c);
+
+                rarrdata[BeamIdx::x  ][pidx] = x;
+                rarrdata[BeamIdx::y  ][pidx] = y;
+                rarrdata[BeamIdx::z  ][pidx] = z;
+                rarrdata[BeamIdx::ux ][pidx] = u[0] * speed_of_light;
+                rarrdata[BeamIdx::uy ][pidx] = u[1] * speed_of_light;
+                rarrdata[BeamIdx::uz ][pidx] = u[2] * speed_of_light;
+                rarrdata[BeamIdx::w  ][pidx] = std::abs(weight);
+
+                // ensure id is always valid
+                // it will repeat if there are more than 2^31-1 particles
+                int id = int((pid + pidx) & ((1ull << 31) - 1));
+                if (id == 0) id = 1;
+
+                iarrdata[BeamIdx::id ][pidx] = id;
+                iarrdata[BeamIdx::cpu][pidx] = 0;
+
```

Comment:
Why can't this be in AddOneBeamParticle? The extra safety for more than 2^31-1 particles should also become relevant for fixed weight beams, so it would be nice to add this to the function and call the function. Or am I missing something?

Reply:
AddOneBeamParticle does not initialize `BeamIdx::cpu` and takes slightly different function arguments than what are available here.